File: 474-ones-and-zeroes/474-ones-and-zeroes.py
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        length = len(strs)
        dp = [[[None for j in range(n + 1)] for i in range(m + 1)] for k in range(length + 1)]
        ans = 0
        freq = [None] * length
        
        for i in range(length):
            string = strs[i]
            zeroes = 0
            ones = 0
            for character in string:
                if character is '0':
                    zeroes += 1
                else:
                    ones += 1
            freq[i] = (zeroes, ones)
                
        def find(i, allowed_zeroes, allowed_ones):
            if i >= length or (allowed_zeroes <= 0 and allowed_ones <= 0):
                return 0
            if dp[i][allowed_zeroes][allowed_ones] is not None:
                return dp[i][allowed_zeroes][allowed_ones]
            zero, one = freq[i]
            dp[i][allowed_zeroes][allowed_ones] = find(i + 1, allowed_zeroes, allowed_ones)
            if zero <= allowed_zeroes and one <= allowed_ones:
                chosen = 1 + find(i + 1, allowed_zeroes - zero, allowed_ones - one)
                dp[i][allowed_zeroes][allowed_ones] = max(dp[i][allowed_zeroes][allowed_ones], chosen)
            return dp[i][allowed_zeroes][allowed_ones]
        
        for i in range(length):
            zero, one = freq[i]
            ans = max(ans, find(i, m, n))
        
        return ans

File: 474-ones-and-zeroes/474-ones-and-zeroes.cpp
struct Tuple{
    int ones;
    int zeroes;
};
class Solution {
    unordered_map<int, Tuple> freq; // index against number of ones and zeroes they contain
    vector<vector<vector<int>>> dp;
    int select(vector<string>& strings, int idx, int m, int n){
        if(idx >= strings.size())
            return 0;
        if(dp[idx][m][n] != -1)
            return dp[idx][m][n];
        
        int ones = freq[idx].ones;
        int zeroes = freq[idx].zeroes;
        
        int include = 0, exclude = 0;
        
        if(ones <= n and zeroes <= m)
            include = select(strings, idx + 1, m - zeroes, n - ones) + 1;
        exclude = select(strings, idx + 1, m, n);
        
        return dp[idx][m][n] = max(include, exclude);
    }
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        int len = strs.size();
        dp.resize(len, vector<vector<int>>(m + 1, vector<int>(n + 1, -1)));
        
        for(int i = 0; i < len; i++){
            string& s = strs[i];
            int zero = 0, one = 0;
            for(char c : s)
                (c == '1') ? one++ : zero++;
            freq[i] = {one, zero};
        }
        return select(strs, 0, m, n);
    }
};